lib/api/bucketPutObjectLock.js

```diff
@@ -0,0 +1,82 @@
+const { waterfall } = require('async');
+const arsenal = require('arsenal');
+
+const errors = arsenal.errors;
+const ObjectLockConfiguration = arsenal.models.ObjectLockConfiguration;
+
+const parseXML = require('../utilities/parseXML');
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const metadata = require('../metadata/wrapper');
+const { metadataValidateBucket } = require('../metadata/metadataUtils');
+const { pushMetric } = require('../utapi/utilities');
+
+/**
+ * Bucket Put Object Lock - Create or update bucket object lock configuration
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+
+function bucketPutObjectLock(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutObjectLock' });
+
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutObjectLock',
+    };
+    return waterfall([
+        next => parseXML(request.post, log, next),
+        (parsedXml, next) => {
+            const lockConfigClass = new ObjectLockConfiguration(parsedXml);
+            // if there was an error getting object lock configuration,
+            // returned configObj will contain 'error' key
+            process.nextTick(() => {
+                const configObj = lockConfigClass.
+                    getValidatedObjectLockConfiguration();
+                return next(configObj.error || null, configObj);
```

Comment:
If there is an error, you're supposed to return a `bucket` object as argument to the callback, not `configObj`, correct? In which case, maybe you could execute `metadataValidateBucket` first so you have the `bucket` object ready for error cases.

Reply:
Hm that's true. I put that part before the `metadataValidateBucket` because AWS returns MalformedXML and InvalidRequest errors before any permissions errors that would be returned by the validation function, but I do need the bucket object.

Reply:
We use this same order of operations for all of the other put config APIs tho, like putBucketReplication and putBucketLifecycle too

Reply:
@jonathan-gramain what do you think would be the best way to handle it?

Reply:
What do the other operations do, they validate the config first, and do not return a `bucket` object in case of error? If so, this is fine to do the same. In doubt I would be consistent with the other operations to avoid surprises.